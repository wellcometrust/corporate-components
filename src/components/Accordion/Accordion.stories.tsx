import React from 'react';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs';

import Accordion, { AccordionItem } from './Accordion';
import Readme from './Accordion.md';

const AccordionExample = () => {
  const content = text('children', 'Accordion item content');
  const title = text('title', 'Heading text default');
  const titleActive = text('titleActive (optional)', 'Heading text active');

  return (
    <>
      <Accordion>
        <AccordionItem
          key="1"
          title={title}
          titleActive={titleActive}
          variant="chevron"
        >
          {content}
        </AccordionItem>
      </Accordion>
      <br />
      <p>Accordion - the siblings closed when one is open</p>
      <Accordion>
        {[1, 2, 3, 4].map(data => (
          <AccordionItem key={`2-${data}`} title={`${title} ${data}`}>
            {`${content} ${data}`}
          </AccordionItem>
        ))}
      </Accordion>
      <br />
      <p>Accordion - stay open until closed</p>
      <Accordion className="cc-accordion--0">
        <AccordionItem
          key="3"
          title="Long title closed consectetur adipiscing elit. Ut pretium pretium tempor. Ut eget imperdiet neque."
          variant="plus"
        >
          {content}
        </AccordionItem>
      </Accordion>
      <Accordion className="cc-accordion--1">
        <AccordionItem key="4" title={title}>
          {content}
        </AccordionItem>
      </Accordion>
      <Accordion className="cc-accordion--last-child">
        <AccordionItem key="5" title={title}>
          {content}
        </AccordionItem>
      </Accordion>
    </>
  );
};

storiesOf('Components|Accordion', module).add('Accordion', AccordionExample, {
  readme: { sidebar: Readme }
});
