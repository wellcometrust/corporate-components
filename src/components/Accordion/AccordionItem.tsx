import React, { MouseEventHandler, useRef } from 'react';
import cx from 'classnames';

import Button from 'Button';

export type AccordionItemProps = {
  active?: boolean;
  children?: React.ReactNode;
  index?: number;
  isEmpty?: boolean;
  isNested?: boolean;
  onClick?: MouseEventHandler;
  title: string;
  titleActive?: string;
  titleAs?: string;
  variant?: 'chevron' | 'plus';
};

export const AccordionItem = ({
  active = false,
  children,
  index,
  isEmpty,
  isNested,
  onClick,
  title,
  titleActive,
  titleAs,
  variant = 'plus'
}: AccordionItemProps) => {
  const content = useRef(null);
  const classNames = cx('cc-accordion-item', {
    'cc-accordion-item--active': active
  });
  const titleClassNames = cx('cc-accordion-item__title', {
    [`cc-accordion-item__title--nested`]: isNested
  });
  const buttonClassNames = cx('cc-accordion__button', {
    [`cc-accordion__button--${variant}`]: variant
  });

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const TitleElement: any = titleAs || 'h3';

  return (
    <>
      {isEmpty ? (
        <div className={classNames}>
          <TitleElement className={titleClassNames}>
            <span className={buttonClassNames}>{title}</span>
          </TitleElement>
        </div>
      ) : (
        <div className={classNames}>
          <TitleElement className={titleClassNames}>
            <Button
              aria-expanded={active}
              className={buttonClassNames}
              icon={variant === 'chevron' ? 'chevronRight' : 'closeSmall'}
              iconPlacementSwitch
              id={`accordion-button-${index}`}
              onClick={onClick}
              variant="unstyled"
            >
              {active && titleActive ? titleActive : title}
            </Button>
          </TitleElement>

          <div
            aria-labelledby={`accordion-button-${index}`}
            className="cc-accordion__content"
            hidden={!active}
            ref={content}
          >
            {children}
          </div>
        </div>
      )}
    </>
  );
};

export default AccordionItem;
