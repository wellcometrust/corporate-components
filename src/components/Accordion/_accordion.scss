// ----------------------------------
// UI Components
// Accordion
// ----------------------------------
// Design System 1.0 Alpha
// 2019-11-11
// ----------------------------------

:root {
  --accordion-button-text-size: calc(12/16 * var(--text-base-size));
}

// .cc-accordion {}

// .cc-accordion-item {}

[class*='cc-accordion--'] {
  border-bottom: 1px solid var(--colour-grey-10);
}

.cc-accordion--0,
.cc-accordion--only-child {
  border-top: 1px solid var(--colour-grey-10);
}

.cc-accordion--last-child,
.cc-accordion--only-child {
  margin-bottom: var(--space-lg) !important;
}

.cc-accordion-item__title {
  margin: 0;
}

.cc-accordion__content {
  padding: var(--space-unit) 0;
}

.cc-accordion__button {
  @extend %btn-base;
  @extend %btn-size-standard;

  .btn__icon {
    height: 1.5rem;
    position: relative;
    width: 1.5rem;
  }

  .btn__icon--right {
    margin-right: 0;
  }

  .cc-accordion-item--active & svg {
    transform: rotate(0deg);
  }
}

.cc-accordion__button--chevron {
  @extend %anchor-base;
  font-size: var(--accordion-button-text-size);
  font-weight: 600;
  text-transform: uppercase;
  transition: color ease 0.4s;
}

.cc-accordion__button--plus {
  //border-top: 1px solid var(--colour-grey-10);
  display: flex;
  font-size: var(--text-base-size);
  font-weight: bold;
  height: auto;
  justify-content: flex-start;
  line-height: var(--body-line-height-sm);
  padding: 8px 0;
  text-align: left;
  width: 100%;

  .cc-accordion-item--active & svg {
    transition: transform 0.2s ease;
  }

  .btn__icon--right {
    color: var(--colour-blue-60);
    margin-left: auto;
  }

  svg {
    position: absolute;
    right: 5px;
  }

  .btn__text {
    padding-right: var(--space-lg);
  }
}

.cc-accordion__button--chevron svg {
  transform: rotate(-90deg);
}

.cc-accordion__button--plus svg {
  transform: rotate(-45deg);
  transition: transform 0.2s ease;
}
