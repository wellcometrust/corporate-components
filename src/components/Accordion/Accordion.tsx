import React, { Children, cloneElement, useState } from 'react';
import cx from 'classnames';

import AccordionItem from './AccordionItem';

type AccordionProps = {
  children?: JSX.Element | JSX.Element[];
  className?: string;
};

export const Accordion = ({ children, className }: AccordionProps) => {
  const [active, setActive] = useState(-1);
  const classNames = cx('cc-accordion', {
    [`${className}`]: className
  });

  // Selects the given item and deselects if it has been clicked again by setting `active` to -1
  // Assumes only one item will be active
  // TODO: add prop to dictate if items can open individually
  const selectItem = (itemNum: number) => {
    const current = active === itemNum ? -1 : itemNum;
    setActive(current);
  };

  const childrenWithProps = Children.map(children, (child, index) =>
    cloneElement(child, {
      // sets the active item in the accordion
      active: index === active,
      index,

      // passes an onClick handler to the child to allow it to set the itself as the active item
      onClick: () => selectItem(index)
    })
  );

  return <div className={classNames}>{childrenWithProps}</div>;
};

export default Accordion;
export { AccordionItem };
