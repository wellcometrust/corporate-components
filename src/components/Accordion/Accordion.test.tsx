import React, { ComponentProps } from 'react';
import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import Accordion, { AccordionItem } from './Accordion';

const setup = (
  accordionProps: Partial<ComponentProps<typeof Accordion>> = {},
  accordionItemProps: Partial<ComponentProps<typeof AccordionItem>> = {}
) => {
  const accordionItemDefaultProps: ComponentProps<typeof AccordionItem> = {
    title: 'Sample title',
    children: 'Sample content',
    ...accordionItemProps
  };

  const accordionDefaultProps: ComponentProps<typeof Accordion> = {
    children: <AccordionItem {...accordionItemDefaultProps} />,
    ...accordionProps
  };

  return render(<Accordion {...accordionDefaultProps} />);
};

describe('<Accordion />', () => {
  it('renders Accordion', () => {
    const { container } = setup({});
    expect(container).toMatchSnapshot();
  });
  it('allows to toggle an accordion item ', () => {
    const onClick = jest.fn();
    setup({}, { onClick });
    expect(screen.getByText('Sample title')).toBeInTheDocument();
    expect(screen.getByText('Sample content')).not.toBeVisible();

    user.click(screen.getByText('Sample title'));

    expect(screen.getByText('Sample content')).toBeVisible();

    user.click(screen.getByText('Sample title'));

    expect(screen.getByText('Sample content')).not.toBeVisible();
  });
});
