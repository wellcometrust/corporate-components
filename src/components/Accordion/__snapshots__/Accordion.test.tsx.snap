// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> renders Accordion 1`] = `
<div>
  <div
    class="cc-accordion"
  >
    <div
      class="cc-accordion-item"
    >
      <h3
        class="cc-accordion-item__title"
      >
        <button
          aria-expanded="false"
          class="cc-accordion__button cc-accordion__button--plus"
          id="accordion-button-0"
          type="button"
        >
          <span
            class="btn__text"
          >
            Sample title
          </span>
          <span
            aria-hidden="true"
            class="icon btn__icon btn__icon--right"
          >
            <svg
              viewBox="0 0 12 12"
            >
              <path
                d="M11.657 1.757L7.414 6l4.243 4.243-1.414 1.414L6 7.414l-4.243 4.243-1.414-1.414L4.586 6 .343 1.757 1.757.343 6 4.586 10.243.343l1.414 1.414z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </h3>
      <div
        aria-labelledby="accordion-button-0"
        class="cc-accordion__content"
        hidden=""
      >
        Sample content
      </div>
    </div>
  </div>
</div>
`;
